cmd/ltkeycloak/from_mattermost.go

```diff
@@ -0,0 +1,286 @@
+	}
+
+	// Use the terraform output terraform host if a manual one is not provided. Useful for development.
+	if keycloakHost == "" {
+		t, err := terraform.New("", *deploymentConfig)
+		if err != nil {
+			return fmt.Errorf("failed to create terraform client: %w", err)
+		}
+		terraformOutput, err := t.Output()
+		if err != nil {
+			return fmt.Errorf("failed to get terraform output: %w", err)
+		}
+		if terraformOutput.KeycloakServer.PublicDNS == "" {
+			return fmt.Errorf("keycloak database cluster not found in terraform output")
+		}
+		keycloakHost = terraformOutput.KeycloakServer.PublicDNS
+	}
+
+	keycloakRealm, err := cmd.Flags().GetString("keycloak-realm")
+	if err != nil {
+		return fmt.Errorf("failed to read flag: %w", err)
+	}
+
+	userPassword, err := cmd.Flags().GetString("set-user-password-to")
+	if err != nil {
+		return fmt.Errorf("failed to read flag: %w", err)
+	}
+
+	keycloakClient := gocloak.NewClient(keycloakHost)
+	ctx := context.Background()
+
+	keycloakToken, err := keycloakClient.LoginAdmin(
+		ctx,
+		deploymentConfig.ExternalAuthProviderSettings.KeycloakAdminUser,
+		deploymentConfig.ExternalAuthProviderSettings.KeycloakAdminPassword,
+		"master", // TODO: Allow specifying the master realm
+	)
+	if err != nil {
+		return fmt.Errorf("failed to refresh keycloak token: %w", err)
+	}
+
+	doneChan := make(chan struct{})
+	workers := runtime.NumCPU()
+	if workers > 4 {
+		workers = 4
+	}
+	usersTxtChan := make(chan string)
+	usersChan := make(chan *model.User)
+
+	wg := sync.WaitGroup{}
+
+	// Worker to write to the users.txt file
+	go userTxtWriter(usersTxtChan, doneChan, userPassword)
+
+	// Workers to run the sync and call the MM and Keycloak APIs
+	for i := 0; i < workers; i++ {
+		workerNumber := i + 1
+
+		go migrateMattermostUsersToKeycloak(&workerConfig{
+			workerNumber:     workerNumber,
+			usersChan:        usersChan,
+			wg:               &wg,
+			mmClient:         mmClient,
+			keycloakClient:   keycloakClient,
+			keycloakToken:    keycloakToken,
+			keycloakRealm:    keycloakRealm,
+			userPassword:     userPassword,
+			deploymentConfig: deploymentConfig,
+			doneChan:         doneChan,
+		})
+	}
+
+	page := 0
+	perPage := 100
+	for {
+		mlog.Info("fetching mattermost users", mlog.Int("page", page), mlog.Int("per_page", perPage))
+		users, _, err := mmClient.GetUsers(ctx, page, perPage, "")
+		if err != nil {
+			return fmt.Errorf("failed to get users from mattermost: %w", err)
+		}
+
+		for _, user := range users {
+			// Skip bots
+			if user.IsBot {
+				continue
+			}
+
+			// Write the user to the users.txt file
+			usersTxtChan <- user.Email
+
+			// Already migrated
+			if user.AuthService == model.UserAuthServiceSaml {
+				continue
+			}
+
+			if !dryRun {
+				wg.Add(1)
+				usersChan <- user
+			} else {
+				mlog.Info("dry-run: would migrate user", mlog.String("username", user.Username))
+			}
+		}
+
+		if len(users) == 0 {
+			break
+		}
+
+		page++
+	}
+
+	wg.Wait()
+	close(doneChan)
```

Comment:
One last bug is to wait until the workers and `userTxtWriter` has actually returned. Refer to the web_conn.go link that I mentioned earlier. Basically it's ensuring that the goroutines are actually exiting before you move on with the code. I will leave it to you if you want to ensure that.